compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1618,21 +1623,88 @@ boolean isAssignableToFiniteType(BType type, BLangLiteral literalExpr) {
         }
 
         BFiniteType expType = (BFiniteType) type;
-        long matchCount = expType.valueSpace.stream().filter(memberLiteral -> {
+        return expType.valueSpace.stream().anyMatch(memberLiteral -> {
             if (((BLangLiteral) memberLiteral).value == null) {
                 return literalExpr.value == null;
             }
-            // Check whether the member literal and the literal that needs to be tested are of same kind and check
-            // the value equality between them.
-            return memberLiteral.getKind().equals(literalExpr.getKind()) &&
-                    ((BLangLiteral) memberLiteral).value.equals(literalExpr.value);
-        }).count();
+            // Check whether the literal that needs to be tested is assignable to any of the member literal in the
+            // value space.
+            return checkLiteralAssignabilityBasedOnType((BLangLiteral) memberLiteral, literalExpr);
+        });
+    }
 
-        // If more than one match means the value space contains ambiguous values.
-        if (matchCount > 1) {
-            dlog.error(literalExpr.pos, DiagnosticCode.AMBIGUOUS_TYPES, type);
+    /**
+     * Method to check the literal assignability based on the types of the literals. For numeric literals the
+     * assignability depends on the equivalency of the literals. If the candidate literal could either be a simple
+     * literal or a constant. In case of a constant, it is assignable to the base literal if and only if both
+     * literals have same type and equivalent values.
+     *
+     * @param baseLiteral      Literal based on which we check the assignability.
+     * @param candidateLiteral Literal to be tested whether it is assignable to the base literal or not.
+     * @return true if assignable; false otherwise.
+     */
+    private boolean checkLiteralAssignabilityBasedOnType(BLangLiteral baseLiteral, BLangLiteral candidateLiteral) {
+        // Different literal kinds.
+        if (baseLiteral.getKind() != candidateLiteral.getKind()) {
+            return false;
         }
-        return matchCount == 1;
+        Object baseValue = baseLiteral.value;
+        Object candidateValue = candidateLiteral.value;
+        int candidateTypeTag = candidateLiteral.type.tag;
+
+        // Numeric literal assignability is based on assignable type and numeric equivalency of values.
+        // If the base numeric literal is,
+        // (1) byte: we can assign byte or a int simple literal (Not an int constant) with the same value.
+        // (2) int: we can assign int literal or int constants with the same value.
+        // (3) float: we can assign int simple literal(Not an int constant) or a float literal/constant with same value.
+        // (4) decimal: we can assign int simple literal or float simple literal (Not int/float constants) or decimal
+        // with the same value.
+        switch (baseLiteral.type.tag) {
```

Comment:
Here is an improvement. Check whether we can simplify this logic by taking out its core validation. IIRC, we have written the same validation for type checker as well. We might able to reuse the same core validation in both places.

Reply:
Ack. I will check on the possibilities of uniforming this logic to decouple value/type assignment and type compatibility checking logic currently we have in the type checker when visiting literals.